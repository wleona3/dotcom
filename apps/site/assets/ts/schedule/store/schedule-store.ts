import {
  Reducer,
  createStore,
  Store,
  applyMiddleware,
  Middleware,
  compose
} from "redux";
import { SelectedOrigin } from "../components/__schedule";
import { DirectionId } from "../../__v3api";
import { getParam, updateParams } from "../../helpers/use-params";

type ModalMode = "schedule" | "origin";

export interface StoreProps {
  selectedDirection: DirectionId;
  scheduleFinderDirection: DirectionId | null;
  scheduleFinderOrigin: SelectedOrigin | null;
  modalOpen: boolean;
  modalMode: ModalMode;
}

// required modalMode, optional scheduleFinderOrigin
type OpenModalPayload = Partial<
  Pick<StoreProps, "modalMode" | "scheduleFinderOrigin">
> &
  Pick<StoreProps, "modalMode">;

export type StoreAction =
  | { type: "INITIALIZE" }
  | { type: "CLOSE_MODAL" }
  | { type: "OPEN_MODAL"; payload: OpenModalPayload }
  | {
      type: "CHANGE_SF_ORIGIN";
      payload: Pick<StoreProps, "scheduleFinderOrigin">;
    }
  | {
      type: "CHANGE_SF_DIRECTION";
      payload: Pick<StoreProps, "scheduleFinderDirection">;
    }
  | {
      type: "CHANGE_DIRECTION";
      payload: Pick<StoreProps, "selectedDirection">;
    };

const parseDirection = (dir: string): DirectionId =>
  parseInt(dir, 10) as DirectionId;

export const initializeStateFromUrl = (state: StoreProps): StoreProps => {
  const updatedState = { ...state };
  const initialDirection = getParam("schedule_direction[direction_id]");
  if (initialDirection) {
    updatedState.selectedDirection = parseDirection(initialDirection);
  }

  const initialScheduleFinderDirection = getParam(
    "schedule_finder[direction_id]"
  );
  const initialScheduleFinderOrigin = getParam("schedule_finder[origin]");
  if (initialScheduleFinderDirection && initialScheduleFinderOrigin) {
    updatedState.scheduleFinderDirection = parseDirection(
      initialScheduleFinderDirection
    );
    updatedState.scheduleFinderOrigin = initialScheduleFinderOrigin;
    // open modal!
    updatedState.modalMode = "schedule";
    updatedState.modalOpen = true;
  } else {
    updatedState.scheduleFinderDirection = null;
    updatedState.scheduleFinderOrigin = null;
  }

  return updatedState;
};

export const scheduleStoreReducer: Reducer<StoreProps, StoreAction> = (
  prevState: StoreProps | undefined = {
    selectedDirection: 0,
    scheduleFinderDirection: null,
    scheduleFinderOrigin: null,
    modalOpen: false,
    modalMode: "schedule"
  },
  action: StoreAction
): StoreProps => {
  if (!prevState) return {} as StoreProps;

  let newState = { ...prevState };

  switch (action.type) {
    case "INITIALIZE":
      newState = initializeStateFromUrl(newState);
      break;
    case "CHANGE_DIRECTION":
      newState.selectedDirection = action.payload.selectedDirection;
      // reset Schedule Finder direction
      newState.scheduleFinderDirection = action.payload.selectedDirection;
      break;
    case "CHANGE_SF_DIRECTION":
      // if modal is open, change to origin selection
      if (newState.modalOpen) {
        newState.scheduleFinderOrigin = null;
        newState.modalMode = "origin";
      }
      newState.scheduleFinderDirection = action.payload.scheduleFinderDirection;
      break;
    case "CHANGE_SF_ORIGIN":
      newState.scheduleFinderOrigin = action.payload.scheduleFinderOrigin;
      // reopen schedule modal
      newState.modalMode = "schedule";
      newState.modalOpen = true;
      break;
    case "OPEN_MODAL":
      newState.modalOpen = true;
      newState.modalMode = action.payload.modalMode;
      // set origin if specified
      if (action.payload.scheduleFinderOrigin) {
        newState.scheduleFinderOrigin = action.payload.scheduleFinderOrigin;
      }
      break;
    case "CLOSE_MODAL":
      newState.modalOpen = false;
      // reset Schedule Finder's direction selection to align with page
      newState.scheduleFinderDirection = newState.selectedDirection;
      break;
    default:
      break;
  }

  return newState;
};

/**
 * Actions
 */
export const closeModal = (): StoreAction => ({ type: "CLOSE_MODAL" });

export const changeScheduleFinderOrigin = (
  scheduleFinderOrigin: SelectedOrigin
): StoreAction => ({
  type: "CHANGE_SF_ORIGIN",
  payload: { scheduleFinderOrigin }
});

export const changeScheduleFinderDirection = (
  scheduleFinderDirection: DirectionId | null
): StoreAction => ({
  type: "CHANGE_SF_DIRECTION",
  payload: { scheduleFinderDirection }
});

export const openOriginModal = (): StoreAction => ({
  type: "OPEN_MODAL",
  payload: { modalMode: "origin" }
});
export const openScheduleModal = (): StoreAction => ({
  type: "OPEN_MODAL",
  payload: { modalMode: "schedule" }
});
export const openScheduleModalWithOrigin = (
  scheduleFinderOrigin: string | null
): StoreAction => ({
  type: "OPEN_MODAL",
  payload: { modalMode: "schedule", scheduleFinderOrigin }
});
export const resetDirection = (
  selectedDirection: DirectionId
): StoreAction => ({
  type: "CHANGE_DIRECTION",
  payload: { selectedDirection }
});

/**
 * Selectors
 */
export const getOrigin = (state: StoreProps): SelectedOrigin | null =>
  state.scheduleFinderOrigin;
export const getPageDirection = (state: StoreProps): DirectionId =>
  state.selectedDirection;
export const getDirection = (state: StoreProps): DirectionId => {
  if (state.scheduleFinderDirection === null) {
    return getPageDirection(state);
  }
  return state.scheduleFinderDirection;
};
export const getModalMode = (state: StoreProps): ModalMode => state.modalMode;
export const getModalOpen = (state: StoreProps): boolean => state.modalOpen;

/**
 * Middleware
 *
 * syncURLMiddleware: watches the state for changes that need to be reflected in
 * the URL parameters
 *
 */
export const syncURLMiddleware: Middleware<
  {},
  StoreProps
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
> = _store => next => action => {
  const result = next(action); // dispatch the action
  let paramsToUpdate: Record<string, string | null> = {};
  if (action.type === "CLOSE_MODAL") {
    paramsToUpdate = {
      "schedule_finder[direction_id]": null,
      "schedule_finder[origin]": null
    };
  } else if (action.type !== "INITIALIZE" && action.payload) {
    const payloadKeys = Object.keys(action.payload);
    if (payloadKeys.includes("selectedDirection")) {
      paramsToUpdate[
        "schedule_direction[direction_id]"
      ] = action.payload.selectedDirection.toString();
    }
    if (payloadKeys.includes("scheduleFinderDirection")) {
      paramsToUpdate[
        "schedule_finder[direction_id]"
      ] = action.payload.scheduleFinderDirection.toString();
    }
    if (payloadKeys.includes("scheduleFinderOrigin")) {
      paramsToUpdate["schedule_finder[origin]"] =
        action.payload.scheduleFinderOrigin;
    }
  }

  if (Object.keys(paramsToUpdate).length) {
    updateParams(paramsToUpdate);
  }

  return result;
};

const exposeStore = (reduxStore: Store): void => {
  /* istanbul ignore next */
  if (window.Cypress) window.store = reduxStore;
};

export const createScheduleStore = (directionId: DirectionId): Store => {
  const middleware = [
    applyMiddleware(syncURLMiddleware),
    /* eslint-disable no-underscore-dangle */
    /* istanbul ignore next */
    ...(window.__REDUX_DEVTOOLS_EXTENSION__
      ? /* istanbul ignore next */
        [window.__REDUX_DEVTOOLS_EXTENSION__()]
      : [])
    /* eslint-enable no-underscore-dangle */
  ];

  const store = createStore(
    scheduleStoreReducer,
    {
      selectedDirection: directionId,
      scheduleFinderDirection: directionId,
      scheduleFinderOrigin: null,
      modalOpen: false,
      modalMode: "schedule"
    } as StoreProps,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    compose(...middleware) as any
  );
  exposeStore(store);
  return store;
};

export const mapStateToProps = (state: StoreProps): StoreProps => ({
  ...state
});

export default mapStateToProps;
