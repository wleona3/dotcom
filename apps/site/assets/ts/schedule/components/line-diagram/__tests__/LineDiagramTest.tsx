import React, { PropsWithChildren } from "react";
import * as swr from "swr";
import { mount, ReactWrapper } from "enzyme";
import { cloneDeep, merge } from "lodash";
import LineDiagramAndStopListPage from "../LineDiagram";
import { EnhancedRoute, RouteType } from "../../../../__v3api";
import {
  LineDiagramStop,
  RoutePatternsByDirection,
  SimpleStop
} from "../../__schedule";
import * as routePatternsByDirection from "../../__tests__/test-data/routePatternsByDirectionData.json";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import outwardLineDiagram from "./lineDiagramData/outward.json"; // not a full line diagram
import simpleLiveData from "./lineDiagramData/live-data.json";
import { fireEvent, render, screen, within } from "@testing-library/react";
import { createScheduleStore } from "../../../store/schedule-store";
import { Provider } from "react-redux";

const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];
let lineDiagramBranchingOut = (outwardLineDiagram as unknown) as LineDiagramStop[];

// Mock useSWR to return fixture data
jest.mock("swr", () => {
  return jest.fn(() => ({ data: simpleLiveData }));
});

const route = {
  type: 3 as RouteType,
  name: "route 1",
  long_name: "route 1 long name",
  color: "F00B42",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
};

lineDiagram.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

lineDiagramBranchingOut.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

let lineDiagramBranchingIn = cloneDeep(lineDiagramBranchingOut).reverse();
const CRroute = merge(cloneDeep(route), { type: 2 as RouteType });
lineDiagramBranchingIn.forEach(({ route_stop }) => {
  route_stop.route = CRroute;
  if (route_stop["is_terminus?"]) {
    route_stop["is_beginning?"] = !route_stop["is_beginning?"];
  }
});

const stops = lineDiagram.map(({ route_stop }) => ({
  name: route_stop.name,
  id: route_stop.id,
  is_closed: false,
  zone: route_stop.zone || null
})) as SimpleStop[];

const directionId = 1;

// redux store/provider
const store = createScheduleStore(0);
function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
  return <Provider store={store}>{children}</Provider>;
}

function renderWithProvider(ui: React.ReactElement) {
  return render(ui, { wrapper: Wrapper });
}

test("LineDiagram renders and matches snapshot", () => {
  let { asFragment } = renderWithProvider(
    <LineDiagramAndStopListPage
      lineDiagram={lineDiagram}
      route={route as EnhancedRoute}
      directionId={directionId}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});

describe("LineDiagram", () => {
  beforeEach(() => {
    renderWithProvider(
      <LineDiagramAndStopListPage
        lineDiagram={lineDiagram}
        route={route as EnhancedRoute}
        directionId={directionId}
      />
    );
  });

  test("includes buttons to open the Schedule Finder modal", () => {
    const { modalOpen: initialModalOpen } = store.getState();
    expect(initialModalOpen).toBe(false);
    const openBtn = screen.getAllByText("View schedule", { exact: false })[0];
    fireEvent.click(openBtn);
    const { modalOpen } = store.getState();
    expect(modalOpen).toBe(true);
  });

  test("can search and filter stops", () => {
    expect(document.querySelector(".m-schedule-diagram--searched")).toBeFalsy();
    const searchInput = screen.getByPlaceholderText("Search for a stop");
    fireEvent.change(searchInput, { target: { value: "S" } });
    expect(
      document.querySelector(".m-schedule-diagram--searched")
    ).toBeTruthy();
    expect(document.querySelector("b.u-highlight")).toBeTruthy();
    expect(document.querySelector("b.u-highlight")!.textContent).toEqual("S");

    fireEvent.change(searchInput, { target: { value: "Impossible" } });
    expect(
      screen.queryByText(
        "Try changing your direction or adjusting your search.",
        { exact: false }
      )
    ).toBeTruthy();
  });
});

test.each`
  type | name
  ${0} | ${"Stations"}
  ${1} | ${"Stations"}
  ${2} | ${"Stations"}
  ${3} | ${"Stops"}
  ${4} | ${"Stops"}
`(
  "LineDiagram names stops or stations for route type $type",
  ({ type, name }) => {
    renderWithProvider(
      <LineDiagramAndStopListPage
        lineDiagram={lineDiagram}
        route={
          {
            ...route,
            type: type as RouteType
          } as EnhancedRoute
        }
        directionId={directionId}
      />
    );
    const heading = screen.getByText(name, { exact: false, selector: "h3" });
    expect(heading).toBeTruthy();
  }
);

test.each`
  type | willPoll
  ${0} | ${true}
  ${1} | ${true}
  ${2} | ${true}
  ${3} | ${true}
  ${4} | ${false}
`(
  "LineDiagram requests live data for most route types: $type",
  ({ type, willPoll }) => {
    // don't mock the return value here,
    // we just want to check if it's called
    const useSWRSpy = jest.spyOn(swr, "default");
    renderWithProvider(
      <LineDiagramAndStopListPage
        lineDiagram={lineDiagram}
        route={
          {
            ...route,
            type: type as RouteType
          } as EnhancedRoute
        }
        directionId={directionId}
      />
    );
    expect(useSWRSpy).toHaveBeenCalled();

    if (willPoll) {
      expect(useSWRSpy).toHaveBeenCalledWith(
        "/schedules/line_api/realtime?id=route-1&direction_id=1",
        expect.any(Function),
        expect.objectContaining({ refreshInterval: expect.any(Number) })
      );
    } else {
      // will not ping the realtime endpoint
      expect(useSWRSpy).toHaveBeenCalledWith(
        null,
        expect.any(Function),
        expect.objectContaining({
          refreshInterval: expect.any(Number)
        })
      );
    }
  }
);
