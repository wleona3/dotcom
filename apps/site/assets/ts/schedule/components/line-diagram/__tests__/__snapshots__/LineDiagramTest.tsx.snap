// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineDiagram renders and matches snapshot 1`] = `
<DocumentFragment>
  <h3
    class="m-schedule-diagram__heading"
  >
    Stops
  </h3>
  <div
    class="m-schedule-diagram__filter c-form__input-container"
  >
    <label
      class="sr-only"
      for="stop-search"
    >
      Search for a stop
    </label>
    <input
      class="c-form__input"
      id="stop-search"
      placeholder="Search for a stop"
      type="text"
      value=""
    />
    <button
      aria-label="search"
      class="c-form__submit-btn"
      type="button"
    >
      <span
        aria-hidden="false"
        class="notranslate c-svg__icon-search-reverse-default"
      >
        SVG
      </span>
    </button>
  </div>
  <div
    class="m-schedule-diagram u-no-crowding-data"
  >
    <div
      class="m-schedule-diagram__vehicle"
      style="top: -25px; left: 7px;"
    >
      <span
        data-animation="false"
        data-html="true"
        data-original-title="<div class=\\"m-schedule-diagram__vehicle-tooltip\\">tooltip for vehicle 1 at stop 1</div>"
        data-placement="right"
        data-selector="true"
        data-toggle="tooltip"
        data-trigger="hover focus"
      >
        <span
          aria-hidden="false"
          class="notranslate m-schedule-diagram__vehicle--icon"
        >
          SVG
        </span>
      </span>
    </div>
    <div
      class="m-schedule-diagram__vehicle"
      style="top: -50px; left: 7px;"
    >
      <span
        data-animation="false"
        data-html="true"
        data-original-title="<div class=\\"m-schedule-diagram__vehicle-tooltip\\">tooltip for vehicle 2 at stop 1</div>"
        data-placement="right"
        data-selector="true"
        data-toggle="tooltip"
        data-trigger="hover focus"
      >
        <span
          aria-hidden="false"
          class="notranslate m-schedule-diagram__vehicle--icon"
        >
          SVG
        </span>
      </span>
    </div>
    <div
      class="m-schedule-diagram__vehicle"
      style="top: -50px; left: 7px;"
    >
      <span
        data-animation="false"
        data-html="true"
        data-original-title="<div class=\\"m-schedule-diagram__vehicle-tooltip\\">tooltip for vehicle 3 at stop 2</div>"
        data-placement="right"
        data-selector="true"
        data-toggle="tooltip"
        data-trigger="hover focus"
      >
        <span
          aria-hidden="false"
          class="notranslate m-schedule-diagram__vehicle--icon"
        >
          SVG
        </span>
      </span>
    </div>
    <svg
      aria-labelledby="diagram-title diagram-desc"
      class="line-diagram-svg bus"
      height="100%"
      role="img"
      style="left: 5px;"
      width="40px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title
        id="diagram-title"
      >
        Line diagram for route 1
      </title>
      <desc
        id="diagram-desc"
      >
        4 stops going undefined to undefined
      </desc>
      <defs>
        <pattern
          height="10"
          id="diagonalHatch"
          patternTransform="rotate(45 0 0)"
          patternUnits="userSpaceOnUse"
          width="10"
        >
          <line
            class="line-diagram-svg__line"
            stroke-width="16px"
            y2="8"
          />
        </pattern>
      </defs>
      <line
        class="line-diagram-svg__line"
        stroke="url(#shuttle)"
        stroke-width="10px"
        x1="11px"
        x2="11px"
        y1="0px"
        y2="0px"
      />
      <line
        class="line-diagram-svg__line"
        stroke-width="10px"
        x1="11px"
        x2="11px"
        y1="0px"
        y2="0px"
      />
      <line
        class="line-diagram-svg__line"
        stroke-width="10px"
        x1="11px"
        x2="11px"
        y1="0px"
        y2="0px"
      />
      <circle
        class="line-diagram-svg__stop"
        cx="11px"
        cy="0px"
        data-shuttle="false"
        r="4px"
      />
      <circle
        class="line-diagram-svg__stop"
        cx="11px"
        cy="0px"
        data-shuttle="false"
        r="4px"
      />
      <circle
        class="line-diagram-svg__stop"
        cx="11px"
        cy="0px"
        data-shuttle="false"
        r="4px"
      />
      <circle
        class="line-diagram-svg__stop"
        cx="11px"
        cy="0px"
        data-shuttle="false"
        r="4px"
      />
    </svg>
    <ol>
      <li
        class="m-schedule-diagram__stop"
        style="padding-left: 36px;"
      >
        <section
          class="m-schedule-diagram__content"
        >
          <header
            class="m-schedule-diagram__stop-heading"
          >
            <h4
              class="m-schedule-diagram__stop-link notranslate"
            >
              <a
                href="/stops/line-origin"
              >
                <span>
                  Origin
                </span>
              </a>
            </h4>
            <div
              class="m-schedule-diagram__features"
            >
              <span
                data-animation="true"
                data-html="false"
                data-original-title="Parking"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
              >
                <i
                  aria-hidden="false"
                  class="fa fa-square-parking notranslate m-schedule-diagram__feature-icon"
                />
              </span>
              <span
                data-animation="true"
                data-html="false"
                data-original-title="Accessible"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
              >
                <span
                  aria-hidden="false"
                  class="notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon"
                >
                  SVG
                </span>
              </span>
            </div>
          </header>
          <div
            class="m-schedule-diagram__stop-details"
          >
            <div
              class="m-schedule-diagram__connections"
            />
            <div
              class="m-schedule-diagram__alert"
            >
              Detour
            </div>
          </div>
          <footer
            class="m-schedule-diagram__footer"
          >
            <button
              class="btn btn-link"
              type="button"
            >
              View schedule
            </button>
          </footer>
        </section>
      </li>
      <li
        class="m-schedule-diagram__stop"
        style="padding-left: 36px;"
      >
        <section
          class="m-schedule-diagram__content"
        >
          <header
            class="m-schedule-diagram__stop-heading"
          >
            <h4
              class="m-schedule-diagram__stop-link notranslate"
            >
              <a
                href="/stops/line-stop1"
              >
                <span
                  aria-hidden="false"
                  class="notranslate c-svg__icon-alerts-triangle"
                >
                  SVG
                </span>
                <span
                  class="sr-only"
                >
                  Service alert or delay
                </span>
                Â 
                <span>
                  Stop 1
                </span>
              </a>
            </h4>
            <div
              class="m-schedule-diagram__features"
            >
              <span
                data-animation="true"
                data-html="false"
                data-original-title="Accessible"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
              >
                <span
                  aria-hidden="false"
                  class="notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon"
                >
                  SVG
                </span>
              </span>
            </div>
          </header>
          <div
            class="m-schedule-diagram__stop-details"
          >
            <div
              class="m-schedule-diagram__connections"
            >
              <a
                data-animation="false"
                data-html="false"
                data-original-title="Orange Line"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
                href="/schedules/Orange/line"
              >
                <span
                  class="m-schedule-diagram__connection"
                >
                  <span
                    aria-hidden="false"
                    class="notranslate c-svg__icon"
                  >
                    SVG
                  </span>
                </span>
              </a>
              <a
                data-animation="false"
                data-html="false"
                data-original-title="Green Line C"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
                href="/schedules/Green-C/line"
              >
                <span
                  class="m-schedule-diagram__connection"
                >
                  <span
                    aria-hidden="false"
                    class="notranslate c-svg__icon"
                  >
                    SVG
                  </span>
                </span>
              </a>
            </div>
            <div
              class="m-schedule-diagram__predictions"
            >
              <div
                class="m-schedule-diagram__prediction notranslate"
              >
                <div>
                  Dest1
                </div>
                <div
                  class="m-schedule-diagram__prediction-time"
                >
                  2   min
                </div>
                <div
                  class="m-schedule-diagram__prediction-crowding"
                >
                  <span
                    aria-hidden="true"
                    class="notranslate c-svg__icon c-icon__crowding c-icon__crowding--crowding_unavailable"
                  >
                    SVG
                  </span>
                </div>
              </div>
              <div
                class="m-schedule-diagram__prediction notranslate"
              >
                <div>
                  Dest2
                </div>
                <div
                  class="m-schedule-diagram__prediction-time"
                >
                  5   min
                </div>
                <div
                  class="m-schedule-diagram__prediction-crowding"
                >
                  <span
                    aria-hidden="true"
                    class="notranslate c-svg__icon c-icon__crowding c-icon__crowding--crowding_unavailable"
                  >
                    SVG
                  </span>
                </div>
              </div>
            </div>
          </div>
          <footer
            class="m-schedule-diagram__footer"
          >
            <button
              class="btn btn-link"
              type="button"
            >
              View schedule
            </button>
          </footer>
        </section>
      </li>
      <li
        class="m-schedule-diagram__stop"
        style="padding-left: 36px;"
      >
        <section
          class="m-schedule-diagram__content"
        >
          <header
            class="m-schedule-diagram__stop-heading"
          >
            <h4
              class="m-schedule-diagram__stop-link notranslate"
            >
              <a
                href="/stops/line-stop2"
              >
                <span
                  aria-hidden="false"
                  class="notranslate c-svg__icon-alerts-triangle"
                >
                  SVG
                </span>
                <span
                  class="sr-only"
                >
                  Service alert or delay
                </span>
                Â 
                <span>
                  Stop 2
                </span>
              </a>
            </h4>
            <div
              class="m-schedule-diagram__features"
            >
              <span
                data-animation="true"
                data-html="false"
                data-original-title="Accessible"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
              >
                <span
                  aria-hidden="false"
                  class="notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon"
                >
                  SVG
                </span>
              </span>
            </div>
          </header>
          <div
            class="m-schedule-diagram__stop-details"
          >
            <div
              class="m-schedule-diagram__connections"
            >
              <a
                data-animation="false"
                data-html="false"
                data-original-title="Route 62"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
                href="/schedules/62/line"
              >
                <span
                  class="c-icon__bus-pill--small m-schedule-diagram__connection u-bg--bus"
                >
                  62
                </span>
              </a>
              <a
                data-animation="false"
                data-html="false"
                data-original-title="Route 67"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
                href="/schedules/67/line"
              >
                <span
                  class="c-icon__bus-pill--small m-schedule-diagram__connection u-bg--bus"
                >
                  67
                </span>
              </a>
            </div>
            <div
              class="m-schedule-diagram__predictions"
            >
              <div
                class="m-schedule-diagram__prediction notranslate"
              >
                <div>
                  Dest2
                </div>
                <div
                  class="m-schedule-diagram__prediction-time"
                >
                  7   min
                </div>
                <div
                  class="m-schedule-diagram__prediction-crowding"
                >
                  <span
                    aria-hidden="true"
                    class="notranslate c-svg__icon c-icon__crowding c-icon__crowding--crowding_unavailable"
                  >
                    SVG
                  </span>
                </div>
              </div>
            </div>
          </div>
          <footer
            class="m-schedule-diagram__footer"
          >
            <button
              class="btn btn-link"
              type="button"
            >
              View schedule
            </button>
          </footer>
        </section>
      </li>
      <li
        class="m-schedule-diagram__stop"
        style="padding-left: 36px;"
      >
        <section
          class="m-schedule-diagram__content"
        >
          <header
            class="m-schedule-diagram__stop-heading"
          >
            <h4
              class="m-schedule-diagram__stop-link notranslate"
            >
              <a
                href="/stops/line-destination"
              >
                <span>
                  Destination
                </span>
              </a>
            </h4>
            <div
              class="m-schedule-diagram__features"
            >
              <span
                data-animation="true"
                data-html="false"
                data-original-title="Parking"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
              >
                <i
                  aria-hidden="false"
                  class="fa fa-square-parking notranslate m-schedule-diagram__feature-icon"
                />
              </span>
              <span
                data-animation="true"
                data-html="false"
                data-original-title="Accessible"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
              >
                <span
                  aria-hidden="false"
                  class="notranslate c-svg__icon-acessible-default m-schedule-diagram__feature-icon"
                >
                  SVG
                </span>
              </span>
            </div>
          </header>
          <div
            class="m-schedule-diagram__stop-details"
          >
            <div
              class="m-schedule-diagram__connections"
            >
              <a
                data-animation="false"
                data-html="false"
                data-original-title="Atlantis"
                data-placement="bottom"
                data-selector="true"
                data-toggle="tooltip"
                data-trigger="hover focus"
                href="/schedules/BOAT-A/line"
              >
                <span
                  class="m-schedule-diagram__connection"
                >
                  <span
                    aria-hidden="false"
                    class="notranslate c-svg__icon-bus-small"
                  >
                    SVG
                  </span>
                </span>
              </a>
            </div>
          </div>
          <footer
            class="m-schedule-diagram__footer"
          >
            <button
              class="btn btn-link"
              type="button"
            >
              View schedule
            </button>
          </footer>
        </section>
      </li>
    </ol>
  </div>
</DocumentFragment>
`;
